### PR TITLE
internal(ex-msg): Add helper functions to improve developer experience and add documentation

### PR DESCRIPTION
**Pull request by mhanley00**
## 🦾 Reason to Be

Prior to release we wanted to add some helper functions to make the developer experience better, and add support for encrypted values so users can send non-string values across to be saved.

## ✅ MR Checklist

- [x] **Properly Titled** The title of this MR is of the form `type(scope): Message goes here`. For example, `fixed(brain): We now fill correctly on google.com`. [More info in the project README](https://gitlab.1password.io/dev/core/core#changelog-entries).
- [ ] **Tests** You’ve written tests that prove your changes work correctly and will catch bugs introduced by future generations.
- [x] **Documentation** If you’re adding or improving a new internal API you’ve properly documented its public-facing components with good example usage.
- [ ] **Follow-on Issues** If this MR is providing stubs that other developers will need to implement, you’ve filed requisite issues for them.
- [ ] **Security Review** Asked for security review if this touches on cryptography, authentication, authorization, changes to logging, input validation or anything else that might impact security?

## 📝 Logging

N/A

## 🤔 Thought Process

- Add English strings for item templates instead of end user needing to provide template number
- Provide helper function to send requests to a specific extension ID
- Provide helper function to encrypt individual values in a save request
- Add documentation with unencrypted and encrypted requests, supported item templates, way for developers to contact us to get their extension added to the allowlist

**TODO:**
- Test save requests for each of our item templates to establish the scope of which item templates we are currently able to support
- Block item templates we do not currently support (ie `identity` for example still needs some work)
- Timeout + error handling
 
## 📋 How To Test

1. Build the extension demo app, follow steps [here](https://www.notion.so/1password/extension-messaging-932b9b9912464aaaaa1daef218ae8cf4#3e6a215767ca44988b34586103fb5930)
2. Checkout core branch `nima/implement-encrypted-value-type` and build b5x (remember to add the demo extension app ID to the list of allowed extensions as detailed in Notion)
3. Load the background app as an extension in Chrome and open the background
4. In the background console, run `window.sendCreateCryptoItem("your-local-b5x-extension-id-here")` > expected behavior is that a new tab should open in Chrome with a Save Dialog and a crypto wallet item with an encrypted recovery seed
5. Next, try following the steps in the README to create and save your own encrypted and unencrypted items in `demo-extension/src/background.ts`. Remember to build again after modifying this file.

## 🐛 Possible Impacts

In the future we may want to add support for fully encrypted requests as well as value-only like we're doing in this MR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,26 +30,167 @@ Response:
...
+	</thead>
+	<tbody>
+	<tr>
+		<td><code>title</code></td> <td>string</td>	<td>The suggested title for the item to be saved.</td>
+	</tr>
+	<tr>
+		<td><code>fields</code></td> <td>array</td> <td>Each object in the array has these properties:
+		<ul>
+			<li><code>autocomplete</code> (string): The type of field to fill.</li>
+			<li><code>value</code> (string or object*): The value to be filled in the field. Objects represent encyrpted values; to encrypt a value, <a href="#encrypted">use the built-in `encryptValue` function</a>.</li>
```
Comment (oliverdunk) [inline README.md]: Typo: encrypted

File: api/src/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  EncryptedValueCodec,
+} from "./index";
+import { CategoryReadable, CategoryUuidDictionary } from "./category";
+import * as base64js from "base64-js";
+
+const OPExtensions = [
+  "dppgmdbiimibapkepcbdbmkaabgiofem",
+  "aeblfdkhhhdcdjpifhhbdiojplfjncoa",
+  "khgocmkkpikpnmmkgmdnfckapcdkgfaf",
+  "{d634138d-c276-4fc8-924b-40a0ea21d284}",
```
Comment (oliverdunk) [inline api/src/helpers.ts:14]: I'd love to see us somehow distinguish between the IDs we use for Chrome, Firefox and Edge and pick the right ones according to the browser we're running in.

File: api/src/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  Response extends Extract<ExtensionResponse, { name: Request["name"] }>["data"]
+>(extensionId: string, message: Request): Promise<Response> {
+  return new Promise<Response>((resolve, reject) => {
+    // TODO add timeout / reject
+    try {
+      chrome.runtime.sendMessage(
+        extensionId,
+        JSON.stringify(message),
+        (response) => {
+          // console.log("Response:", response);
```
Comment (oliverdunk) [inline api/src/helpers.ts]: Can we drop this line entirely?

File: api/src/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+          resolve(response);
+        }
+      );
+    } catch {
+      reject("not found");
```
Comment (oliverdunk) [inline api/src/helpers.ts]: Can we have a more descriptive error here? Also curious if we've seen any "unhandled chrome.runtime.lastError" messages testing this, if so we may need to check that property so Chrome knows not to log it: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/lastError

File: api/src/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      type: "encrypted",
+      value: base64js.fromByteArray(new Uint8Array(encryptResult)),
+    };
+  } catch {
+    throw new Error("Unable to encrypt value provided");
+  }
+};
+
+export async function createOPItem(
+  extensionId: string,
```
Comment (oliverdunk) [inline api/src/helpers.ts:44]: Personally I'd still love to see some sort of extension class that stores the ID so we don't need to pass an extension ID in to every call. I'm not sure if that fits in to this MR but happy to chat more if you want to know what I'm thinking.

File: api/src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  if (CategoryUuidDictionary[itemType]) {
+    const categoryUuid = CategoryUuidDictionary[itemType];
+    const response = (await sendExternalMessage(extensionId, {
+      name: "create-item",
+      data: {
+        saveRequest,
+        type: categoryUuid,
+      },
+    })) as unknown;
+    return response as t.TypeOf<typeof CreateItemResponseDataCodec>;
```
Comment (oliverdunk) [inline api/src/helpers.ts]: Have we tried typing the `sendExternalMessage` so we don't need this cast?

**Comment by oliverdunk** [comment]
Left some initial comments! Overall this is looking great though, super excited to be providing more code for people 🙌

**Review by pudgeball — APPROVED** [review]

File: api/src/helpers.ts (hunk 2 above)
Comment (pudgeball) [reply api/src/helpers.ts:14]: Extracting to an issue in Core

File: api/src/helpers.ts (hunk 4 above)
Comment (pudgeball) [reply api/src/helpers.ts]: Extracting to an issue in Core

File: api/src/helpers.ts (hunk 5 above)
Comment (pudgeball) [reply api/src/helpers.ts:44]: Extracting to an issue in Core

**Pull request merged**